uvicorn/supervisors/statreload.py

```diff
@@ -24,15 +24,23 @@ def __init__(self, config, target, sockets):
         self.should_exit = threading.Event()
         self.pid = os.getpid()
         self.mtimes = {}
+        self.child_pids = []
 
     def signal_handler(self, sig, frame):
         """
         A signal handler that is registered with the parent process.
         """
+        for child_pid in self.child_pids:
+            try:
+                os.kill(child_pid, signal.SIGINT)
+                finished = os.waitpid(child_pid, 0)
+            except Exception as e:
+                logger.error(f"Cant kill child PID {child_pid}: {e}")
```

Comment:
Any chance this logic can be factored out to a utility, eg `shutdown_children(pids)`? Or maybe even a `ProcessTracker` helper class that both supervisors would use. Suggesting this because the implementation here and in the multi process supervisor are basically duplicates.

Reply:
I did it, still pass the tests on the docker containers, not sure about the super() calls etc, it always hurts my brain

Reply:
Hmm. Okay, the refactor doesn't seem right - we shouldn't need any subclassing - and I actually don't think we need any other abstraction than a `shutdown_subprocess()` helper.

Examining the existing source code, I also think we don't even need a new `child_pids` list — multiprocess has the list of `processes`, and reloaders store the reload `process`, so we can use those directly.

Here's the diff patch I ended up with working through this locally, WDYT?

```diff
diff --git a/uvicorn/subprocess.py b/uvicorn/subprocess.py
index 735a513..f583db2 100644
--- a/uvicorn/subprocess.py
+++ b/uvicorn/subprocess.py
@@ -5,6 +5,7 @@ starting child processes.
 import multiprocessing
 import os
+import signal
 import sys
 
 multiprocessing.allow_connection_pickling()
 spawn = multiprocessing.get_context("spawn")
@@ -59,3 +60,13 @@ def subprocess_started(config, target, sockets, stdin_fileno):
 
     # Now we can call into `Server.run(sockets=sockets)`
     target(sockets=sockets)
+
+
+def shutdown_subprocess(pid):
+    """
+    Helper to attempt cleanly shutting down a subprocess. May fail with an exception.
+
+    * pid - Process identifier.
+    """
+    os.kill(pid, signal.SIGINT)
+    os.waitpid(pid, 0)
diff --git a/uvicorn/supervisors/basereload.py b/uvicorn/supervisors/basereload.py
index 7dbe4dd..21b82c0 100644
--- a/uvicorn/supervisors/basereload.py
+++ b/uvicorn/supervisors/basereload.py
@@ -5,7 +5,7 @@ import threading
 
 import click
 
-from uvicorn.subprocess import get_subprocess
+from uvicorn.subprocess import get_subprocess, shutdown_subprocess
 
 HANDLED_SIGNALS = (
     signal.SIGINT,  # Unix signal 2. Sent by Ctrl+C.
@@ -27,6 +27,11 @@ class BaseReload:
         """
         A signal handler that is registered with the parent process.
         """
+        try:
+            shutdown_subprocess(self.process.pid)
+        except Exception as exc:
+            logger.error(f"Could not stop reload process {self.process.pid}: {exc}")
+
         self.should_exit.set()
 
     def run(self):
diff --git a/uvicorn/supervisors/multiprocess.py b/uvicorn/supervisors/multiprocess.py
index 94f7238..2dce1bc 100644
--- a/uvicorn/supervisors/multiprocess.py
+++ b/uvicorn/supervisors/multiprocess.py
@@ -5,7 +5,7 @@ import threading
 
 import click
 
-from uvicorn.subprocess import get_subprocess
+from uvicorn.subprocess import get_subprocess, shutdown_subprocess
 
 HANDLED_SIGNALS = (
     signal.SIGINT,  # Unix signal 2. Sent by Ctrl+C.
@@ -28,6 +28,12 @@ class Multiprocess:
         """
         A signal handler that is registered with the parent process.
         """
+        for process in self.processes:
+            try:
+                shutdown_subprocess(process.pid)
+            except Exception as exc:
+                logger.error(f"Could not stop child process {process.pid}: {exc}")
+
         self.should_exit.set()
 
     def run(self):
```

You can copy this and `git apply file.patch` to get the changes. :)

Reply:
I'm sorry @florimondmanca but I really think maintaing child pids is key in the above issue
I'll rebase off master and put a rather annoying to run test that wont pass with the above patch but succeeds with changes I propose